server/src/main/java/com/cloud/network/router/VirtualNetworkApplianceManagerImpl.java

```diff
@@ -2578,7 +2579,13 @@ protected void finalizeNetworkRulesForNetwork(final Commands cmds, final DomainR
                 }
             }
 
-            final List<LoadBalancerVO> lbs = _loadBalancerDao.listByNetworkIdAndScheme(guestNetworkId, Scheme.Public);
+            List<LoadBalancerVO> lbs = null;
+            Long vpcId = guestNetwork.getVpcId();
+            if (vpcId != null) {
+                lbs = _loadBalancerDao.listByVpcIdAndScheme(vpcId, Scheme.Public);
```

Comment:
I'm not sure if I understood exactly what you meant; since LB rules are being searched for by VPC ID, exceptions should not occur if there are missing NICs. In the worst case, there will be a rule added to VR of one network that does not exists. 

Regarding applying the rules when all guest networks are added to VR, from what I saw of the current ACS workflow, the solution I presented corrects the problem, and I didn't find another part of the code where it can be solved besides this section. If you think that this solution could be applied to another section, could you please point to where that would be?

Reply:
@SadiJr 
the method `finalizeNetworkRulesForNetwork` is called for each guest network when VPC VR is started

https://github.com/apache/cloudstack/blob/706f3f6ab429be888f982c72e08689b99b9a428e/server/src/main/java/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java#L507-L523

If you search for `StartCommand` in management server logs, you will see multiple `LoadBalancerConfigCommand` with same parameters.

Reply:
Correct, the method is indeed called for each guest network, but the `LoadBalancerConfigCommand` command, from what I've analyzed in the logs, is not the same. Instead, for each guest network that calls such a method, only the load balancer rules for that network are passed in the `LoadBalancerConfigCommand` command. This is the reason why, in a VPC with multiple guest networks, each of which has its own load balancer rules, only the rules of the last implemented network are applied.

Reply:
I know what caused the issue and why you made the changes.
What I meant is, the process can be optimized. Only one LoadBalancerConfigCommand (with LB of all tiers) is needed, maybe after all other rules are applied.

Reply:
I can understand your point of view, however, considering the criticality and the time it would take me to implement and test your proposal, I'll continue with the current implementation and create an issue to optimize such a process in the future, using your idea. What do you think?

Reply:
@SadiJr 
if you do not mind, I can create a new PR for it tomorrow

Reply:
Sure, feel free to create an alternative PR, as it allows the community to compare the solutions and choose the best one.

Reply:
> @SadiJr if you do not mind, I can create a new PR for it tomorrow

@weizhouapache , would this be an alternative or an addition?

Have you created it?